modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/ranges.scala

```diff
@@ -0,0 +1,67 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.data.NonEmptyList
+import cats.syntax.all._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.node.shared.domain.seedlist.Range.ValidationResult
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+
+case class InvalidInput(str: String) extends NoStackTrace {
+  override def getMessage = s"Invalid input: $str"
+}
+
+case object NoValidRanges extends NoStackTrace
+
+object ranges {
+
+  private val delimiter = ";"
+
+  def decode(input: String): ValidationResult[NonEmptyList[Range]] =
+    input
+      .split(delimiter)
+      .toList
+      .traverse(Range.decode)
+      .andThen {
+        validate(_)
+          .leftMap(_ => InvalidInput(input))
+          .toValidatedNel
+      }
+      .map(mergeOverlaps)
+
+  private def mergeOverlaps(ranges: NonEmptyList[Range]): NonEmptyList[Range] = {
+    val merged = ranges
+      .sortBy(_.start.value)
+      .foldLeft(List.empty[Range]) {
+        case (prev :: tail, curr) if curr.start - prev.end <= 1 =>
+          val merged = Range(
+            prev.start,
+            NonNegLong.unsafeFrom(
+              Math.max(prev.end, curr.end)
+            )
+          )
+
+          merged :: tail
+        case (acc, curr) =>
+          curr :: acc
+      }
+      .reverse
+
+    NonEmptyList(merged.head, merged.tail)
```

Comment:
It's better to explicitly declare you use unsafe function, instead of using `.head` which is kinda hidden and no so obvious at the first glance. You don't need to split `.head` and `.tail`, just use `NonEmptyList.fromListUnsafe(merged)`

Reply:
Changed to:

```scala
  private def mergeOverlaps(ranges: NonEmptyList[Range]): NonEmptyList[Range] = {
    val merged = ranges
      .sortBy(_.start.value)
      .foldLeft(List.empty[Range]) {
        case (prev :: tail, curr) if curr.start - prev.end <= 1 =>
          val merged = Range(
            prev.start,
            NonNegLong.unsafeFrom(
              Math.max(prev.end, curr.end)
            )
          )

          merged :: tail
        case (acc, curr) =>
          curr :: acc
      }
      .reverse

    NonEmptyList.fromListUnsafe(merged)
  }
```